src/main/sensors/gyro.h

```diff
@@ -59,30 +59,46 @@ typedef enum {
 #define GYRO_CONFIG_USE_GYRO_2      1
 #define GYRO_CONFIG_USE_GYRO_BOTH   2
 
+typedef enum {
+    FILTER_LOWPASS = 0,
+    FILTER_LOWPASS2
+} filterSlots;
+
 typedef struct gyroConfig_s {
     sensor_align_e gyro_align;              // gyro alignment
     uint8_t  gyroMovementCalibrationThreshold; // people keep forgetting that moving model while init results in wrong gyro offsets. and then they never reset gyro. so this is now on by default.
     uint8_t  gyro_sync_denom;                  // Gyro sample divider
     uint8_t  gyro_lpf;                         // gyro LPF setting - values are driver specific, in case of invalid number, a reasonable default ~30-40HZ is chosen.
-    uint8_t  gyro_soft_lpf_type;
-    uint8_t  gyro_soft_lpf_hz;
+
     bool     gyro_high_fsr;
     bool     gyro_use_32khz;
     uint8_t  gyro_to_use;
-    uint16_t gyro_soft_lpf2_hz;
+
+    // Lagged Moving Average smoother
+    uint8_t gyro_lma_depth;
+    uint8_t gyro_lma_weight;
+
+    // Lowpass primary/secondary
+    uint8_t  gyro_lowpass_type;
+    uint8_t  gyro_lowpass2_type;
+
+    // Order is used for the 'higher ordering' of cascaded butterworth/biquad sections
+    uint8_t  gyro_lowpass_order;
+    uint8_t  gyro_lowpass2_order;
+
+    uint16_t gyro_lowpass_hz;
+    uint16_t  gyro_lowpass2_hz;
+
     uint16_t gyro_soft_notch_hz_1;
     uint16_t gyro_soft_notch_cutoff_1;
     uint16_t gyro_soft_notch_hz_2;
     uint16_t gyro_soft_notch_cutoff_2;
     gyroOverflowCheck_e checkOverflow;
-    uint16_t gyro_filter_q;
-    uint16_t gyro_filter_r;
-    uint16_t gyro_filter_p;
     int16_t  gyro_offset_yaw;
-    uint8_t  gyro_soft_lpf2_order;
+
 } gyroConfig_t;
 
-#define GYRO_LPF2_ORDER_MAX 6
```

Comment:
It was intentional to separate maximum supported butterworth size BIQUAD_LPF_ORDER_MAX, limited by number of Q coefficients, and `GYRO_LPF2_ORDER_MAX `, limited by allocated memory

Reply:
Makes sense. Re-introduced. The only problem I see is the structure in filter.h uses `BIQUAD_LPF_ORDER_MAX` as the `GYRO_LPF_ORDER_MAX` there is not reachable. So potentially would have to re-work the structure array allocation to be in gyro.c/.h and use `GYRO_LPF_ORDER_MAX` 

Let me know what you think the best way to handle is.

Reply:
The idea was that filter.c function simply expects that caller provided large enough array/caller takes care of allocation.
It was easy when counter was stored in separate place ..

One (a bit obscure way) is to store section count in function pointer - biquadCascadeFilterApply1, biquadCascadeFilterApply2, biquadCascadeFilterApply3 (each calling   generic Apply function with count parameter) and return correct function pointer during biquad initialization.

Or declarare 0-section type in filter.h (gcc `-pedantic` will complain) and use larger structure in gyro.c

Or provide filter type as parametric macro in filter.c:
```
#define BIQUAD_CASCADE_STATE(order) struct {  int sections; biquadFilter_t biquad[((order) + 1) / 2]; }
```
With `void*` state parameter to apply function, it will even suppress need for casting ... 

---

It would probably be easiest to use full filter.h declaration for now - it can be solved later (and filtering needs rewrite anyway, this problem may simply vanish )